### ISSUE TITLE
Unsupported CPU: 0x50650 Intel(R) Xeon(R) Silver 4114

### ISSUE THREAD

**Issue by ntrrgc** [mozilla/rr#2144]
rr is not working on my CPU:

```
[FATAL /home/roc/rr/rr/src/PerfCounters.cc:193:get_cpu_microarch() errno: EOPNOTSUPP] CPU 0x50650 unknown.
=== Start rr backtrace:
rr(_ZN2rr13dump_rr_stackEv+0x28)[0x554cb8]
rr(_ZN2rr15notifying_abortEv+0x47)[0x554d37]
rr[0x4b6796]
rr(_ZN2rr12PerfCountersC1Ei+0x35)[0x4b7165]
rr(_ZN2rr4TaskC2ERNS_7SessionEiijNS_13SupportedArchE+0x5c)[0x53be8c]
rr(_ZN2rr10RecordTaskC1ERNS_13RecordSessionEijNS_13SupportedArchE+0x24)[0x4f7214]
rr(_ZN2rr13RecordSession8new_taskEiijNS_13SupportedArchE+0x3b)[0x4bd09b]
rr(_ZN2rr4Task5spawnERNS_7SessionERKNS_8ScopedFdERKNS_11TraceStreamERKSsRKSt6vectorISsSaISsEESF_i+0x181)[0x53fae1]
rr(_ZN2rr13RecordSessionC1ERKSsRKSt6vectorISsSaISsEES7_NS0_16SyscallBufferingENS_7BindCPUE+0x1ad)[0x4c4cdd]
rr(_ZN2rr13RecordSession6createERKSt6vectorISsSaISsEES5_NS0_16SyscallBufferingENS_7BindCPUE+0x7d5)[0x4c5655]
rr(_ZN2rr13RecordCommand3runERSt6vectorISsSaISsEE+0x944)[0x4bc164]
rr(main+0x244)[0x44f2e4]
/lib64/libc.so.6(__libc_start_main+0xea)[0x7f7ad379303a]
rr[0x44f7cf]
=== End rr backtrace
Aborted (core dumped)
```

lscpu output:
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  10
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
Stepping:            4
CPU MHz:             800.000
CPU max MHz:         2201,0000
CPU min MHz:         800,0000
BogoMIPS:            4400.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            14080K
NUMA node0 CPU(s):   0-9,20-29
NUMA node1 CPU(s):   10-19,30-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 intel_ppin intel_pt mba tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke
```

**Comment by khuey** [comment]
Looks like a shiny new processor that we don't know about.  Add `0x50650` to the `IntelSkylake` cases at https://github.com/mozilla/rr/blob/master/src/PerfCounters.cc#L170 and verify that `make check` passes on your machine?

**Comment by ntrrgc** [comment]
Basic debugging seems to work, but not all tests pass. These are the results:

```
99% tests passed, 12 tests failed out of 2161

Total Test time (real) = 297.73 sec

The following tests FAILED:
	724 - diversion_syscall (Failed)
	725 - diversion_syscall-no-syscallbuf (Failed)
	936 - call_exit (Failed)
	937 - call_exit-no-syscallbuf (Failed)
	1072 - unwind_on_signal (Failed)
	1073 - unwind_on_signal-no-syscallbuf (Failed)
	1804 - diversion_syscall-32 (Failed)
	1805 - diversion_syscall-32-no-syscallbuf (Failed)
	2016 - call_exit-32 (Failed)
	2017 - call_exit-32-no-syscallbuf (Failed)
	2152 - unwind_on_signal-32 (Failed)
	2153 - unwind_on_signal-32-no-syscallbuf (Failed)
Errors while running CTest
make[3]: *** [CMakeFiles/check.dir/build.make:57: CMakeFiles/check] Error 8
make[2]: *** [CMakeFiles/Makefile2:68: CMakeFiles/check.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:75: CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:197: check] Error 2
```

**Comment by rocallahan** [comment]
Can you dump the full output of `ctest -j8 --verbose` somewhere we can read it? And what OS are you on? Thanks!!! It does look like rr should basically work for you.

**Comment by rocallahan** [comment]
I pushed https://github.com/mozilla/rr/commit/747fbbd3ac1f6cb477d46cd0f9c2dccc5d42721d to support this Skylake model.

**Comment by ntrrgc** [comment]
rr works on simple examples but crashes when debugging WebKit, whereas I don't have that problem in my laptop with a different CPU.

```
*** Error in `/webkit/WebKitBuild/Debug/bin/WebKitWebProcess': double free or corruption (fasttop): 0x00007fbd2c4adee0 ***
[New Thread 34140.34157]
[New Thread 34140.34162]
[New Thread 34140.34165]
[New Thread 34140.34168]
[New Thread 34140.34171]
[New Thread 34140.34175]
[New Thread 34140.34178]
[New Thread 34140.34181]
[New Thread 34140.34246]
[New Thread 34140.34249]
[New Thread 34140.34267]
[New Thread 34140.34283]
[New Thread 34140.34304]
[New Thread 34140.34307]
[New Thread 34140.34310]
[New Thread 34140.34313]
[New Thread 34140.34317]
[New Thread 34140.34320]

Thread 4 received signal SIGABRT, Aborted.
[Switching to Thread 34140.34259]
0x00007fbdc578569b in raise () from /lib64/libc.so.6
(rr) b free
Breakpoint 1 at 0x7fbdc57e1280 (7 locations)
(rr) reverse-continue 
Continuing.
[FATAL /home/ntrrgc/Apps/rr/src/EmuFs.cc:100:create() errno: EMFILE] Failed to create shmem segment /tmp/rr-emufs-34375-dev-64768-inode-31992919-\home\ntrrgc\.cache\fontconfig\1e7d33d4977304646dae68e658eb3ad2-le64.cache-7
=== Start rr backtrace:
rr(_ZN2rr13dump_rr_stackEv+0x28)[0x54c9f8]
rr(_ZN2rr15notifying_abortEv+0x47)[0x54ca77]
rr(_ZN2rr7EmuFile6createERNS_5EmuFsERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmmm+0x991)[0x45bc11]
rr(_ZN2rr7EmuFile5cloneERNS_5EmuFsE+0x78)[0x45c078]
rr(_ZN2rr5EmuFs10clone_fileESt10shared_ptrINS_7EmuFileEE+0x22)[0x45c202]
rr(_ZN2rr7Session13copy_state_toERS0_RNS_5EmuFsES3_+0x1225)[0x5274c5]
rr(_ZN2rr13ReplaySession5cloneEv+0x128)[0x4fa768]
rr(_ZN2rr14ReplayTimeline18seek_to_before_keyERKNS0_7MarkKeyE+0x144)[0x511034]
rr(_ZN2rr14ReplayTimeline16reverse_continueERKSt8functionIFbPNS_10ReplayTaskEEERKS1_IFbvEE+0x15a2)[0x516332]
rr(_ZN2rr9GdbServer14debug_one_stepERNS_10GdbRequestE+0x807)[0x485e47]
rr(_ZN2rr9GdbServer12serve_replayERKNS0_15ConnectionFlagsE+0x68b)[0x48777b]
rr[0x4f5060]
rr(_ZN2rr13ReplayCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE+0x561)[0x4f6181]
rr(main+0x1e8)[0x554bad]
/lib64/libc.so.6(__libc_start_main+0xea)[0x7f2de00cf03a]
rr(_start+0x2a)[0x43d2aa]
=== End rr backtrace
Remote connection closed
```

ctest output: [ctest.txt](https://github.com/mozilla/rr/files/1507760/ctest.txt)

**Comment by khuey** [comment]
All of the failing tests have something of the form

`'open' has unknown return type; cast the call to its declared return type`

in their logs. That's probably a gdb or libc symbols issue, not something with rr itself.

**Comment by ntrrgc** [comment]
This is Fedora 27 (also in my laptop).

gdb is custom built though because the one provided with the distro makes gdbserver crash. https://sourceware.org/bugzilla/show_bug.cgi?id=22137

```
GNU gdb (GDB) 8.0.50.20171013-git
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
```

**Comment by rocallahan** [comment]
````
[FATAL /home/ntrrgc/Apps/rr/src/EmuFs.cc:100:create() errno: EMFILE] Failed to create shmem segment /tmp/rr-emufs-34375-dev-64768-inode-31992919-\home\ntrrgc\.cache\fontconfig\1e7d33d4977304646dae68e658eb3ad2-le64.cache-7
````

Looks like your fd limit was exhausted. Not sure which version of rr you're on, but rr master tries to increase your fd limit to 64K. So if you're not on rr master, try that.

If you are on rr master, try increasing your fd limit (via `ulimit`) to 64K and make sure that works (and see if it helps). Or possibly even a higher limit...

**Comment by ntrrgc** [comment]
That output if from master.

If having too many file descriptors was the cause, shouldn't the same happen in my laptop which has the same limits?

**Comment by rocallahan** [comment]
Probably, but it's possible that fds are consumed at a different rate on different machines, depending on configuration details. `EMFILE` indicates out of fds, at least.

Can you try manually raising the fd limit to something ridiculous like 1M and see what happens?

**Comment by ntrrgc** [comment]
Indeed. Raising the fd limit worked. Turns out rr opens a lot of file descriptors. In this machine and this program it was about ~6000, which was above the 4096 limit.

Regarding the error `'open' has unknown return type; cast the call to its declared return type`, it seems to be related to my specific version of gdb and it was fixed by making it find system library symbols. First I fixed it for a single session with:

```
show debug-file-directory # show the previous one, to confirm that it's not the system (expected one)
set debug-file-directory /usr/lib/debug
```

Then I made it permanent by rebuilding with `--with-separate-debug-dir=/usr/lib/debug`.

8 tests still failed, but they should be a problem with this gdb, as using system one removes all the failures. I'll close the issue.
